src/crl.c

```diff
@@ -490,27 +490,204 @@ int BufferLoadCRL(WOLFSSL_CRL* crl, const byte* buff, long sz, int type,
 }
 
 #if defined(OPENSSL_EXTRA) && defined(HAVE_CRL)
+/* helper function to create a new dynamic WOLFSSL_X509_CRL structure */
+static WOLFSSL_X509_CRL* wolfSSL_X509_crl_new(WOLFSSL_CERT_MANAGER* cm)
+{
+    WOLFSSL_X509_CRL* ret;
+
+    ret = (WOLFSSL_X509_CRL*)XMALLOC(sizeof(WOLFSSL_X509_CRL), cm->heap,
+                DYNAMIC_TYPE_CRL);
+    if (ret != NULL) {
+        if (InitCRL(ret, cm) < 0) {
+            WOLFSSL_MSG("Unable to initialize new CRL structure");
+            XFREE(ret, cm->heap, DYNAMIC_TYPE_CRL);
+            ret = NULL;
+        }
+    }
+    return ret;
+}
+
+
+/* returns head of copied list that was alloc'd */
+static RevokedCert *DupRevokedCertList(RevokedCert* in, void* heap)
+{
+    RevokedCert* head = NULL;
+    RevokedCert* current = in;
+    RevokedCert* prev = NULL;
+    while (current) {
+        RevokedCert* tmp = (RevokedCert*)XMALLOC(sizeof(RevokedCert), heap,
+                DYNAMIC_TYPE_REVOKED);
+        if (tmp != NULL) {
+            XMEMCPY(tmp->serialNumber, current->serialNumber,
+                    EXTERNAL_SERIAL_SIZE);
+            tmp->serialSz = current->serialSz;
+            tmp->next = NULL;
+            if (prev != NULL)
+                prev->next = tmp;
+            if (head == NULL)
+                head = tmp;
+        }
+        current = current->next;
+    }
+    return head;
+}
+
+
+/* returns a deep copy of ent on success and null on fail */
+static CRL_Entry* DupCRL_Entry(CRL_Entry* ent, void* heap)
+{
+    CRL_Entry *dup;
+
+    dup = (CRL_Entry*)XMALLOC(sizeof(CRL_Entry), heap, DYNAMIC_TYPE_CRL_ENTRY);
+    if (dup == NULL) {
+        WOLFSSL_MSG("alloc CRL Entry failed");
+        return NULL;
+    }
+
+    XMEMCPY(dup->issuerHash, ent->issuerHash, CRL_DIGEST_SIZE);
+    XMEMCPY(dup->lastDate, ent->lastDate, MAX_DATE_SIZE);
+    XMEMCPY(dup->nextDate, ent->nextDate, MAX_DATE_SIZE);
+    dup->lastDateFormat = ent->lastDateFormat;
+    dup->nextDateFormat = ent->nextDateFormat;
+    dup->certs = DupRevokedCertList(ent->certs, heap);
+
+    dup->totalCerts = ent->totalCerts;
+    dup->verified = ent->verified;
+
+    if (!ent->verified) {
+        dup->tbsSz = ent->tbsSz;
+        dup->signatureSz = ent->signatureSz;
+        dup->signatureOID = ent->signatureOID;
+        dup->toBeSigned = (byte*)XMALLOC(dup->tbsSz, heap,
+                                          DYNAMIC_TYPE_CRL_ENTRY);
+        if (dup->toBeSigned == NULL) {
+            XFREE(dup, heap, DYNAMIC_TYPE_CRL_ENTRY);
```

Comment:
Fail case would leak `dup->certs`. Is there a CRL cleanup function that can be used?